packages/block-editor/src/components/colors-gradients/panel-color-gradient-settings.js

```diff
@@ -0,0 +1,152 @@
+	return settings.map( ( {
+		colorValue,
+		gradientValue,
+		label,
+		colors: availableColors,
+		gradients: availableGradients,
+	}, index ) => {
+		if ( ! colorValue && ! gradientValue ) {
+			return null;
+		}
+		let ariaLabel;
+		if ( colorValue ) {
+			const colorObject = getColorObjectByColorValue(
+				availableColors || colors,
+				colorValue
+			);
+			ariaLabel = sprintf(
+				colorIndicatorAriaLabel,
+				label.toLowerCase(),
+				( colorObject && colorObject.name ) || colorValue
+			);
+		} else {
+			const gradientObject = __experimentalGetGradientObjectByGradientValue(
+				availableGradients || gradients,
+				colorValue
+			);
+			ariaLabel = sprintf(
+				gradientIndicatorAriaLabel,
+				label.toLowerCase(),
+				( gradientObject && gradientObject.name ) || gradientValue
+			);
+		}
+
+		return (
+			<ColorIndicator
+				key={ index }
+				colorValue={ colorValue || gradientValue }
+				aria-label={ ariaLabel }
+			/>
+		);
+	} );
+};
+
+export const PanelColorGradientSettingsInner = ( {
+	className,
+	colors,
+	gradients,
+	disableCustomColors,
+	disableCustomGradients,
+	children,
+	settings,
+	title,
+	...props
+} ) => {
+	if (
+		isEmpty( colors ) &&
+		isEmpty( gradients ) &&
+		disableCustomColors &&
+		disableCustomGradients &&
+		every(
+			settings,
+			( setting ) => (
+				isEmpty( setting.colors ) &&
+				isEmpty( setting.gradients ) &&
+				( setting.disableCustomColors === undefined || setting.disableCustomColors ) &&
+				( setting.disableCustomGradients === undefined || setting.disableCustomGradients )
+			)
+		)
+	) {
+		return null;
+	}
+
+	const titleElement = (
+		<span className="block-editor-panel-color-gradient-settings__panel-title">
+			{ title }
+			<Indicators colors={ colors } gradients={ gradients } settings={ settings } />
+		</span>
+	);
+	return (
+		<PanelBody
+			className={ classnames( 'block-editor-panel-color-gradient-settings', className ) }
+			title={ titleElement }
+			{ ...props }
+		>
+			{ settings.map( ( setting, index ) => (
+				<ColorGradientControl
+					key={ index }
+					{ ...{
+						colors,
+						gradients,
+						disableCustomColors,
+						disableCustomGradients,
+						...setting,
+					} }
+				/>
+			) ) }
+			{ children }
+		</PanelBody>
+	);
+};
+
+const PanelColorGradientSettingsSelect = ( props ) => {
+	const colorGradientSettings = useSelect(
+		( select ) => {
+			const settings = select( 'core/block-editor' ).getSettings();
+			return pick( settings, colorsAndGradientKeys );
+		}
+	);
+	return <PanelColorGradientSettingsInner { ...{ ...colorGradientSettings, ...props } } />;
+};
+
+const PanelColorGradientSettings = ( props ) => {
+	const relevantProps = pick( props, colorsAndGradientKeys );
+	if ( isEmpty( relevantProps ) || some(
+		pick( props, colorsAndGradientKeys ),
+		( setting ) => ( setting === undefined )
+	) ) {
+		return <PanelColorGradientSettingsSelect { ...props } />;
+	}
+	return <PanelColorGradientSettingsInner { ...props } />;
+};
```

Comment:
Just observing that this is the same mechanics as in `control.js`. Maybe it speaks to a lack of higher-level tooling in our environment (e.g. `isEmpty( foo ) || some( foo, isUndefined )`, plus the fact that (presumably?) this _if_ statement is meant to avoid calling `useSelect` in some cases but not others).

Reply:
I was making this logic unnecessarily complex, we can simply use `every( colorsAndGradientKeys, ( key ) => ( props[ key ] !== undefined ) )`.